### DISCUSSION TITLE
Define `useFetch` behavior

### DISCUSSION THREAD

**Discussion by userquin** [vueuse/vueuse#575] [General]
@wheatjs @antfu @benlesh @isgj 

Can we do a little effort to close the `useFetch` functioning? I know we are all very busy with other projects, but this subject is taking more than 10 days to just only define its behavior (we have disparate behaviors on #549, I'm suggesting using raw objects instead fetch ones, others requesting removing the raw object encoding and also the content-type header handling on multipart requests)...

I can remove all logic to encode raw objects to build the request body and also remove the content-type header handling and just keep the original logic, only adding the content type when necessary.
The stopper here was the payloadType option since is forcing us to deal with it, but removing the content-type header that handles the issue it goes away.

Problems to be resolved:
- using raw objects or just using `fetch` ones: you can see some examples [here](https://github.com/vueuse/vueuse/pull/549/files#diff-8e2a59a63d7197173044c5bdc03f0c6b63a72b72f7231b855e3c3302d1300da5R165) and [here](https://github.com/vueuse/vueuse/pull/549/files#diff-8e2a59a63d7197173044c5bdc03f0c6b63a72b72f7231b855e3c3302d1300da5R225). Raw objects can be used to change the encoding without rewriting the logic: for example, switching between `application/json` and `application/x-www-form-urlencoded` is just switch the `payloadType` option without rewriting the logic to encode the object while using just `fetch` ones we will need to change also the object itself we are sending.
- try to correct invalid options from user options: deal with multipart header if present. You can see the problem [here](https://github.com/vueuse/vueuse/pull/549#discussion_r644682493).
- access to the payload on error: see proposal on #549 for `responseHandler` and `CustomResponseContext` type [here](https://github.com/vueuse/vueuse/pull/549/files#diff-0db62ed388d474bc80cd69e2ff1593cd6c9719b8794d8f88459a1704e8da25b5R158) and [here](https://github.com/vueuse/vueuse/pull/549/files#diff-0db62ed388d474bc80cd69e2ff1593cd6c9719b8794d8f88459a1704e8da25b5R112) respectivelly.

**Comment by wheatjs** [comment]
I'm currently working on a PR for this. I've been looking at how other popular libraries have done things and am going to follow what they do. Currently following what `use-http` does as they have already addressed many of the issues we are currently having. https://github.com/ava/use-http [[see their implementation for handling headers here](https://github.com/ava/use-http/blob/master/src/doFetchArgs.ts)]

**Reply by benlesh** [reply]
If I might make a suggestion: it's a noble goal to try to make useFetch cover everyone's use cases, however, I think it's in the best interest of the ecosystem to make as many of the features as tree-shakeable as possible. A collection of composition-based utilities. Even the useHttp example is a bit heavy, IMHO. People should only pay for what they use. 

Sorry for butting my head in. This isn't my library. Just an area of interest for me.

What you're working on here could well be one of the most important and most used utilities in the Vue ecosystem. The weight of the design is as important as the features, for sure.

**Reply by wheatjs** [reply]
> If I might make a suggestion: it's a noble goal to try to make useFetch cover everyone's use cases, however, I think it's in the best interest of the ecosystem to make as many of the features as tree-shakeable as possible. A collection of composition-based utilities. Even the useHttp example is a bit heavy, IMHO. People should only pay for what they use.
> 
> Sorry for butting my head in. This isn't my library. Just an area of interest for me.
> 
> What you're working on here could well be one of the most important and most used utilities in the Vue ecosystem. The weight of the design is as important as the features, for sure.

I don't mind at all and do appreciate the input. When I was mentioning doing what `useHttp` does, I mostly just meant the headers and body part. Do you think that would be too much? Please feel free to speak your mind, any input is greatly appreciated

**Comment by antfu** [comment]
I personally would like to keep this function as simple as possible. It's more like, if you want to fetch something quickly, we already provided out-of-box. For more complex cases, they should use `axios`, `ky` or whatever they want with more mature and well-rounded implementations. Otherwise, a separate package like `@vueuse/fetch` would be better, but I would also doubt if it's worth to do given there is already many lib doing great jobs on this.

**Reply by wheatjs** [reply]
Since it looks like we would all like to keep `useFetch` pretty simple, perhaps we can remove the portion of `useFetch` that sets the request content-type and modifies the post data. This would end up being a breaking change, but might be the best option as we can avoid giving the wrong idea of this being a replacement for something like axios and we can keep the function minimal. I'm not not a big fan of having something like `@vueuse/fetch` as I think we are all pretty busy and it would be just another thing to maintain.

**Reply by isgj** [reply]
Yesterday I made a PR #573 

It sets the content type only if no content type is given and the payload is `object`. The API doesn't change and it will work also with other types for which before the content type was set as `text`.

It handles only the case `json` so also the payload is stringified because (probably) it's the most used case. For the rest it proxies to  `fetch`

**Comment by mariusa** [comment]
As a newcomer, I thought `useFetch` was a replacement for `axios`. Would you please clarify the intended use case and make some recommendations at https://vueuse.org/core/usefetch/ ?

Also, coming from `axios`, I was expecting the possibility to use `await`, eg
```
let response = await axios.get('/api/auth/session')
console.log(response.data) //when reaching this line, data is set. Errors are handled in interceptor. Headers are set in interceptor.
```

How can one use `useFetch` in a similar, simple way? None of the given examples use `await`, so how does one simply get the data without additional code?

**Reply by userquin** [reply]
@mariusa `useFetch` uses the `reactivity` system (`composition api`) and so you don't need to use `await` since it will be done by the `useFetch` function, you get a `ref` to the response body, along other `ref`s to know if the `useFetch` is fetching, has finished, has error...:

```ts
const {
  data, // <= body payload ref
  error,
  abort,
  statusCode,
  isFetching, // <= fetching ref
  isFinished, // <= fetch finish ref
  canAbort,
  execute,
} = useFetch(url, { refetch }).get()
```

You can see the `demo.vue` file on this repo or link from footer page on docs.

**Reply by mariusa** [reply]
Thanks. Let's consider the use case that one wants to process retrieved data, right after retrieval:
```
const {
  data, // <= body payload ref
  error,
  abort,
  statusCode,
  isFetching, // <= fetching ref
  isFinished, // <= fetch finish ref
  canAbort,
  execute,
} = useFetch(url, { refetch }).get()

for (let i=0; i<data.rows.length; i++) {
   data.rows[i].email = data.rows[i].email.toLowerCase()
}
```
This won't work, as data isn't set yet (?). One has to wait for fetch to complete.

How should one handle this? Have continuous a loop to check if isFinished is set? (ugly)

**Reply by userquin** [reply]
`useFetch` is the composable hook for `setup`, that will register all stuff, and so you only need to add a `watch` for `data ref` (since it is a `ref` you need to use `data.value` (on `watch` there is no need):

```ts
watch(data, (newData, oldData) => {
  // newData contains the plain data body from fetch
}) 
```

In the original example, it is fetching async (is feching asap), since `immediate` is `true` by default (I'm not 100% sure), you will get the response once the `fetch` finish.

You can see a playground [here](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbiAgaW1wb3J0IHsgcmVmLCBjb21wdXRlZCB9IGZyb20gJ3Z1ZSdcblx0aW1wb3J0IHsgdXNlRmV0Y2ggfSBmcm9tICdAdnVldXNlL2NvcmUnXG4gIGltcG9ydCB7IHVzZVRvZ2dsZSB9IGZyb20gJ0B2dWV1c2Uvc2hhcmVkJ1xuICBcbiAgY29uc3QgbmF0aXZlRGF0YSA9IHJlZignJylcbiAgY29uc3QgdXNlTmF0aXZlID0gcmVmKGZhbHNlKVxuIFxuICBcbiAgY29uc3QgYnVpbGRGb3JtRGF0YSA9ICgpID0+IHtcbiAgICBjb25zdCBmb3JtRGF0YSA9IG5ldyBGb3JtRGF0YSgpXG4gICAgZm9ybURhdGEuYXBwZW5kKFxuICAgICAgXCJBXCIsIFxuICAgICAgbmV3IEJsb2IoXG4gICAgICAgIFtKU09OLnN0cmluZ2lmeSh7IG1lc3NhZ2U6IGBNdWx0aXBhcnQgQTogJHtNYXRoLnJhbmRvbSgpfWB9KV0sIFxuICAgICAgICB7IHR5cGU6ICdhcHBsaWNhdGlvbi9qc29uJyB9XG4gICAgICApLCBcbiAgICAgICd0ZXN0LW11bHRpcGFydC1hLmpzb24nXG4gICAgKVxuICAgIGZvcm1EYXRhLmFwcGVuZChcbiAgICAgIFwiQlwiLCBcbiAgICAgIG5ldyBCbG9iKFxuICAgICAgICBbSlNPTi5zdHJpbmdpZnkoeyBtZXNzYWdlOiBgTXVsdGlwYXJ0IEI6ICR7TWF0aC5yYW5kb20oKX1gfSldLCBcbiAgICAgICAgeyB0eXBlOiAnYXBwbGljYXRpb24vanNvbicgfVxuICAgICAgKSwgXG4gICAgICAndGVzdC1tdWx0aXBhcnQtYi5qc29uJ1xuICAgIClcbiAgICByZXR1cm4gZm9ybURhdGFcbiAgfVxuXG4gIGNvbnN0IHsgLypwb3N0LCAqL2V4ZWN1dGUsIGRhdGEgfSA9IHVzZUZldGNoKCdodHRwczovL2h0dHBiaW4ub3JnL3Bvc3QnLCB7XG4gICAgaW1tZWRpYXRlOiBmYWxzZVxuICB9KS5wb3N0KGJ1aWxkRm9ybURhdGEoKSwgJ2Zvcm1EYXRhJylcbiAgXG4gIGNvbnN0IHVzZUZldGNoRGF0YSA9IGNvbXB1dGVkKCgpID0+IHtcbiAgICBpZiAobmF0aXZlRGF0YS52YWx1ZSkge1xuICAgICAgcmV0dXJuICcnXG4gICAgfSBlbHNlIHtcbiAgICAgIGlmIChkYXRhLnZhbHVlKSB7XG4gICAgICAgIHJldHVybiBKU09OLnN0cmluZ2lmeShkYXRhLnZhbHVlKVxuICAgICAgfSBlbHNlIHtcbiAgICAgICAgcmV0dXJuICcnXG4gICAgICB9XG4gICAgfVxuICB9KVxuICBcbiAgY29uc3QgdXNpbmdVc2VGZXRjaCA9IGFzeW5jKCkgPT4ge1xuICAgIC8vY29uc3QgcGF5bG9hZCA9IGF3YWl0IGJ1aWxkRm9ybURhdGEoKVxuICAgIC8vY29uc29sZS5sb2cocG9zdClcbiAgICBuYXRpdmVEYXRhLnZhbHVlID0gJydcbiAgICB1c2VOYXRpdmUudmFsdWUgPSBmYWxzZVxuICAgIHRyeSB7XG4gICAgICAvL2NvbnNvbGUubG9nKHBvc3QocGF5bG9hZCwgJ2Zvcm1EYXRhJykpXG4gICAgICBhd2FpdCBleGVjdXRlKClcbiAgICB9IGNhdGNoKGUpIHtcbiAgICAgIC8vIGp1c3QgaWdub3JlXG4gICAgICBjb25zb2xlLmxvZygndXBwcycsIGUpXG4gICAgfVxuICB9XG4gIGNvbnN0IHVzaW5nTmF0aXZlRmV0Y2ggPSBhc3luYygpID0+IHtcbiAgICBuYXRpdmVEYXRhLnZhbHVlID0gJydcbiAgICB1c2VOYXRpdmUudmFsdWUgPSB0cnVlXG4gICAgdHJ5IHtcblx0XHRcdGNvbnN0IHJlc3BvbnNlID0gYXdhaXQgZmV0Y2goJ2h0dHBzOi8vaHR0cGJpbi5vcmcvcG9zdCcsIHtcbiAgICAgICAgbWV0aG9kOiAnUE9TVCcsXG4gIFx0XHRcdGJvZHk6IGJ1aWxkRm9ybURhdGEoKSxcbiAgICAgIH0pXG4gICAgICBuYXRpdmVEYXRhLnZhbHVlID0gYXdhaXQgcmVzcG9uc2UuanNvbigpXG4gICAgfSBjYXRjaCAoaWdub3JlKSB7XG4gICAgICAvLyBqdXN0IGlnbm9yZSBlcnJvclxuICAgIH1cbiAgfVxuPC9zY3JpcHQ+XG48dGVtcGxhdGU+XG4gIDxkaXYgY2xhc3M9XCJjb250YWluZXJcIj5cbiAgICA8ZGl2IGNsYXNzPVwiYnV0dG9uc1wiPlxuICAgICAgPGJ1dHRvbiBAY2xpY2s9XCJ1c2luZ1VzZUZldGNoXCI+XG4gICAgICAgIG11bHRpcGFydCB1c2luZyA8c3Ryb25nPnVzZUZldGNoPC9zdHJvbmc+XG4gICAgICA8L2J1dHRvbj5cbiAgICAgIDxidXR0b24gQGNsaWNrPVwidXNpbmdOYXRpdmVGZXRjaFwiPlxuICAgICAgICBtdWx0aXBhcnQgdXNpbmcgPHN0cm9uZz5uYXRpdmUgZmV0Y2g8L3N0cm9uZz5cbiAgICAgIDwvYnV0dG9uPlxuICAgIDwvZGl2PlxuICAgIDxkaXY+XG4gICAgICBDaGVjayA8c3Ryb25nPm11bHRpcGFydC9mb3JtLWRhdGE7IGJvdW5kYXJ5IG9uIENvbnRlbnQtVHlwZSByZXF1ZXN0IGhlYWRlcjwvc3Ryb25nPiBvbiBkZXZ0b29sc1xuICAgIDwvZGl2PlxuICAgIDxkaXYgY2xhc3M9XCJkYXRhXCI+XG4gICAgICA8cHJlIHYtaWY9XCJ1c2VOYXRpdmVcIj5cbiAgICAgICAge3sgbmF0aXZlRGF0YSB9fVxuICAgICAgPC9wcmU+XG4gICAgICA8cHJlIHYtZWxzZT5cbiAgICAgICAge3sgdXNlRmV0Y2hEYXRhIH19XG4gICAgICA8L3ByZT5cbiAgICA8L2Rpdj5cbiAgIFxuICA8L2Rpdj5cbjwvdGVtcGxhdGU+XG48c3R5bGU+XG4gIC5jb250YWluZXIge1xuICAgIGRpc3BsYXk6IGdyaWQ7XG4gICAgZ3JpZC10ZW1wbGF0ZS1jb2x1bW5zOiAxZnI7XG4gICAgZ3JpZC1yb3ctZ2FwOiAxcmVtO1xuICAgIHBhZGRpbmc6IDFyZW07XG4gICAganVzdGlmeS1pdGVtczogY2VudGVyO1xuICB9XG4gIC5idXR0b25zIHtcbiAgICBkaXNwbGF5OiBncmlkO1xuICAgIGdyaWQtdGVtcGxhdGUtY29sdW1uczogcmVwZWF0KDIsIDFmcik7XG4gICAgZ3JpZC1jb2x1bW4tZ2FwOiAxcmVtO1xuICB9XG4gIC5kYXRhIHtcbiAgICB3aWR0aDogMTAwJTtcbiAgICBvdmVyZmxvdy14OiBhdXRvO1xuICB9XG4gIHByZSB7XG4gICAgZm9udC1zaXplOiAwLjZyZW07XG4gIH1cbjwvc3R5bGU+IiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuXHRcdFwiQHZ1ZXVzZS9jb3JlXCI6IFwiaHR0cHM6Ly91bnBrZy5jb20vQHZ1ZXVzZS9jb3JlL2luZGV4LmVzbS5qc1wiLFxuICAgIFwiQHZ1ZXVzZS9zaGFyZWRcIjogXCJodHRwczovL3VucGtnLmNvbS9AdnVldXNlL3NoYXJlZC9pbmRleC5lc20uanNcIixcbiAgICBcInZ1ZS1kZW1pXCI6IFwiaHR0cHM6Ly91bnBrZy5jb20vdnVlLWRlbWkvbGliL2luZGV4LmVzbS5qc1wiXG4gIH1cbn0ifQ==)

**Reply by userquin** [reply]
In your example, move the `for loop` inside the `watch` and use `newData` instead `data`.

**Reply by userquin** [reply]
the data type on data `ref` may be `text`,  `json`,  `blob`... see the docs.

**Reply by mariusa** [reply]
I see, thanks for clarification! It makes sense, but it's more complex than `await`. I'll stick to `axios`.

**Reply by isgj** [reply]
@manaust  in your case the bare `fetch` is the option.

`useFetch` was explained before and it's use case, but in my use cases when I wanted to do something after the call was finished I did it like so

```js
const { isFetching, data, execute } = useFetch('/', { immediate: false }) // you can add other config like .jaon()

// when it makes sense to make the call and await
await execute() // if you use POST method you can do: await post({...}).execute()
myProccessData(data) // here you have access to data and all the rest
```

if you don't need to show the state of the call(or handle it manually) just use `axios` or `fetch`, in my case I wanted to show loading state and await for the call

**Reply by isgj** [reply]
((oops I mentioned the wrong user))
